File: python/snake_game/snake.py
"""snake body, movement, growth, collisions. grid cells only, no pygame import"""
from collections import deque

UP = (0, -1)
DOWN = (0, 1)
LEFT = (-1, 0)
RIGHT = (1, 0)


def opposite(direction):
    dx, dy = direction
    return (-dx, -dy)


class Snake:
    def __init__(self, start_cell, direction=RIGHT):
        self.body = deque([start_cell])
        self.direction = direction
        self._pending_direction = direction
        self._growth_pending = 0

    @property
    def head(self):
        return self.body[0]

    @property
    def tail(self):
        return self.body[-1]

    @property
    def length(self):
        return len(self.body)

    def set_direction(self, direction):
        # can't reverse into own neck if longer than 1
        if self.length > 1 and direction == opposite(self.direction):
            return
        self._pending_direction = direction

    def grow(self, amount=1):
        self._growth_pending += amount

    def peek_next_head(self):
        dx, dy = self._pending_direction
        hx, hy = self.head
        return (hx + dx, hy + dy)

    def advance(self):
        """steps one cell, applies growth, returns new head"""
        self.direction = self._pending_direction
        new_head = self.peek_next_head()
        self.body.appendleft(new_head)
        if self._growth_pending > 0:
            self._growth_pending -= 1
        else:
            self.body.pop()
        return new_head

    def hits_wall(self, cols, rows):
        x, y = self.head
        return x < 0 or x >= cols or y < 0 or y >= rows

    def hits_self(self):
        head = self.head
        body_iter = iter(self.body)
        next(body_iter)  # skip the head itself
        return head in body_iter

    def occupied_cells(self):
        return set(self.body)

File: cpp/src/snake.cpp
#include "snake.hpp"

#include <algorithm>

Snake::Snake(Cell start, Direction direction)
    : direction_(direction), pendingDirection_(direction) {
    body_.push_back(start);
}

void Snake::SetDirection(Direction direction) {
    if (Length() > 1 && direction == Opposite(direction_)) {
        return;
    }
    pendingDirection_ = direction;
}

void Snake::Grow(int amount) { growthPending_ += amount; }

Cell Snake::PeekNextHead() const {
    Cell delta = DirectionDelta(pendingDirection_);
    const Cell& head = Head();
    return Cell{head.col + delta.col, head.row + delta.row};
}

Cell Snake::Advance() {
    direction_ = pendingDirection_;
    Cell newHead = PeekNextHead();
    body_.push_front(newHead);
    if (growthPending_ > 0) {
        growthPending_--;
    } else {
        body_.pop_back();
    }
    return newHead;
}

bool Snake::HitsWall(int cols, int rows) const {
    const Cell& head = Head();
    return head.col < 0 || head.col >= cols || head.row < 0 || head.row >= rows;
}

bool Snake::HitsSelf() const {
    const Cell& head = Head();
    return std::find(body_.begin() + 1, body_.end(), head) != body_.end();
}
